File: Civil_Engineering/Misc/FEM_Python/Linear/getForceFromGravity.py
import numpy as np
from elements import Element
from getElementCoordinates import getElementCoordinates
def getForceFromGravity(NodalCoord, Connectivity, rho, g, h,el_type='Q4',P_type='vector'):
    if (el_type == 'Q4'):
        num_node = 4  # nodes per element
    elif (el_type == 'Q8'):
        num_node = 8
    else:
        print("Please enter a valid element type")

    # now lets get the degrees of freedom (i.e. num_dof)
    if (P_type == 'vector'):
        num_dof = num_node * 2  # elemental dof
        num_dof_total = 2 * len(NodalCoord)
        A = np.zeros((len(Connectivity), num_dof))  # create assembly matrix which is num_elements x num_dof
        for e in range(len(Connectivity)):  # for each element
            for i in range(num_node):  # all the columns of a row
                for j in range(2):  # num of dof per node
                    if j == 0:
                        A[e, (i) * 2 + j] = 2 * Connectivity[e][i] - 1  # assign x first based of nodal id of certain element hence
                        # con_mat[e,i]
                    else:
                        A[e, (i) * 2 + j] = 2 * Connectivity[e][i]  # assign y from nodal ID

    else:
        print("Please enter a valid problem type")

    F = np.zeros((num_dof_total,1))  # K matrix will be num_dof x num_dof

    for e in range(0, len(Connectivity)):
        #Coord_mat_el = getElementCoordinates(e, NodalCoord, Connectivity)
        # Loop over each element
        fe = F_GRAV([[0],[-rho*g]])*h
        for i in range(num_dof):
            dof_1 = int(A[e, i])-1 # get degrees of freedom and subtract 1 for now due to ordering
            F[dof_1, 0] += fe[i,0]

    # end elemental loop
    return F


def F_GRAV(trac):
    f_grav = np.zeros((8,1))
    E = Element("Q4")

    weights = [1, 1]
    xIP = [-np.sqrt(3) / 3, np.sqrt(3) / 3]
    for k in range(len(xIP)):
        s = xIP[k];
        N = E.N(s,s)
        J_s = np.matrix(detJsideQ4(s,s))
        n_mat = np.matrix(N.transpose())
        trac_mat =  np.matrix(trac)
        f_grav += n_mat*trac_mat*J_s[0,k]*weights[k]
    return f_grav

def detJsideQ4(x,y):
    G = np.array([-1/2,1/2])
    dxdpsi = G*x
    dydpsi = G*y
    detJside = np.sqrt(dxdpsi**2+dydpsi**2)
    return detJside

File: Civil_Engineering/530_FEM/HW8/Matlab_routines/getForceFromGravity.m
function [Fg] = getForceFromGravity(NodalCoord, Connectivity, rho, g, thickness)
%This functions computes element force vectors from a gravity body force
%and assembles the contributions into a global vector.
%We want to apply a force of -pg on every y coordinate and nothing on the
%x.
num_node = 4;
    num_dof = num_node*2;
    A = zeros(length(NodalCoord),num_dof); %create assembly matrix which is num_elements x num_dof
    for e = 1:2 %each element
         for i =1:num_node %all the columns of a row
             for j = 1:2 %num of dof per node
                % disp(j)
                  if (j==1)
                      A(e,(i-1)*2+j) = 2*Connectivity(e,i)-1 ; %x Dof
                    %disp(j)
                  else
                        A(e,(i-1)*2+j) = 2*Connectivity(e,i); %y Dof
                  end
             end %end j
         end %end i
    end%end e
    
 Fg = zeros(12,1);
 % now to actually apply the load!
    function f_grav = F_GRAV(trac)
        f_grav = zeros(8,1); %set initial
        %traction = zeros(8,1);
        %for i = 1:8
        %    current_coord = Assembly(element,i);%get nodal value
        %    if (mod(current_coord,2) == 0)
        %        traction(i) = trac(2);
        %    else
        %        traction(i) = 0;
        %    end
        %oend %filled traction
        %time to integrate
        X_ip =  [-1/sqrt(3),1/sqrt(3)];
        W_ip = [1,1];
        for k = 1:2
           s = X_ip(k);
           %disp(s)
           Ne = Shape_func(s,s);
           J_s = detJSideQ4(s,s);
           
           f_grav = f_grav + transpose(Ne)*trac*J_s(k)*W_ip(k);
        end
        
    end
 
 
for e = 1:2
   % Coords_el = getElementCoordinates(e, NodalCoord, Connectivity);
    fe = F_GRAV([0;-rho*g])*thickness;
    %print(K_e)

    %print(Coord_mat_el.T) %check that proper values are being returned (they are)

    for i= 1:8
        dof_1 = A(e,i);  % get degrees of freedom
        Fg(dof_1) = Fg(dof_1) +  fe(i);
    %end i loop
    end
%end elemental loop
end


end

